### ISSUE TITLE
Transaction Conflict

### ISSUE THREAD

**Issue by careri** [Starcounter/Starcounter#3192]
Is there any way of enabling more logging for why a conflict occurs? I'm trying to reproduce my conflict with less code that I could publish here. But I can't reproduct with a simple example.

And I can't figure out why I get a conflict. So is there any way of getting more info?

**Comment by per-samuelsson** [comment]
@careri 

Good question. I don't think there are, but others probably know this better.

/cc @eriohl @k-rus @linkdata @chrhol

**Comment by eriohl** [comment]
No there is not I'm afraid. What specifically are you looking for? Id of record(s) in conflict? What do you intend to use it for?

**Comment by per-samuelsson** [comment]
> What do you intend to use it for?

I'm guessing mainly like a way to diagnose a problem, considering this:

> And I can't figure out why I get a conflict.

**Comment by careri** [comment]
Yes, it would be awsome  you could enable some slower diagnostic with ID and field ID. It could even be some special diagnostics build of the kernel.

I can reproduce all the time now. First attempt fails, second is successful. I'll try writing some own diagnostics tomorrow.

**Comment by eriohl** [comment]
Conflicts are detected on record level, not field level. We do not track which fields has changed. So maybe some misconception on this is why you can't figure out why you get a conflict?

What I meant was why you worried about why you get a conflict in this particular case? Generally they are almost impossible to avoid completely (unless you are only running one transaction at a time) and in general the only realistic thing you can to about it it to redo the transaction.

But it is possible get the record id(s) of the records in conflict in the error message (as done for index constraint violation). Not in the near future however (where are pretty occupied at the moment).

If you need more immediate help we can perhaps do a special build with a log of something as you suggest?

**Comment by careri** [comment]
Morning!

The thing is that I have one http service that creates one item-record in a Db.Transact for each call to the service. The item-record has a sequence number (unique combined index) in combination with it's head-record. When the record is created the sequence number counter on the head-record is increased and stored on the item-record.

I then have a consumer that reads (no write) item-records and once they are considered processed I delete them.

So one provider transaction when the item is created. And one consumer transaction that reads and deletes.

I can't understand how I can get a conflict on that? Is it the sequence number counter on the head-record that is the problem? I don't touch the counter in the consumer transaction.

Could it have something to do with isolation level, I know that there is something called snapshot. But I haven't read up on what it does.

I'll put some more effort into reproducing in a simpler app.

**Comment by careri** [comment]
Here's code that will reproduce:
```
using Starcounter;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;

namespace ScConflict
{
    class Program
    {
        static Program()
        {
            new DebugMonitor(s_source);
        }

        private static readonly CancellationTokenSource s_source = new CancellationTokenSource();
        private static readonly AutoResetEvent s_available = new AutoResetEvent(false);
        private static readonly Random s_random = new Random();

        static void Main()
        {
            // Create unique index
            try
            {
                using (var t = new Transaction())
                {
                    Db.SQL("CREATE UNIQUE INDEX dummy2_uid ON ScConflict.DummyLog2 (Head Asc, SortId Asc)");
                }
            }
            catch { }

            Db.Transact(DummyLog2.Purge);
            Db.Transact(DummyHead.Purge);
            Hook<DummyLog2>.CommitInsert += OnInsertOrUpdate;
            Hook<DummyLog2>.CommitUpdate += OnInsertOrUpdate;
            Db.DatabaseStopping += OnCancel;
            var dbSession = new DbSession();
            dbSession.RunAsync(ProvideLogs, 0);
            dbSession.RunAsync(ProvideLogs, 1);
            dbSession.RunAsync(ConsumeLogs);
        }

        static void OnInsertOrUpdate(object sender, DummyLog2 e)
        {
            s_available.Set();
        }

        private static void OnCancel(object sender, EventArgs e)
        {
            s_source.Cancel();
        }
            
        private static void ProvideLogs()
        {
            int tid = Thread.CurrentThread.ManagedThreadId;
            int counter = 0;
            DummyHead head = null;
            Db.Transact(() => head = new DummyHead());

            while (!s_source.IsCancellationRequested)
            {
                try
                {
                    for (int i = 0; i < s_random.Next(1, 10); i++)
                    {
                        Db.Transact(() => new DummyLog2(head, string.Format("{0}.{1}.{2}", tid, counter, i)));
                    }
                }
                catch (Exception ex)
                {
                    WriteLog("Failed to provide logs: " + ex);
                }
                counter++;
                Thread.Sleep(s_random.Next(1000, 5000));
            }
        }

        private static void ConsumeLogs()
        {
            // Try to leave the enumerator open and find new items (doesn't work)
            var waitHandles = new WaitHandle[] { s_available, s_source.Token.WaitHandle };
            var maxItems = 10;
            var maxTime = TimeSpan.FromSeconds(5);

            while (!s_source.IsCancellationRequested)
            {
                if (WaitHandle.WaitAny(waitHandles) == 0)
                {
                    int attempt = -1;

                    // Create a batch
                    Db.Transact(() =>
                        {
                            attempt++;
                            var itemsSet = new HashSet<DummyLog2>();
                            var stopWatch = Stopwatch.StartNew();

                            while (itemsSet.Count < maxItems && stopWatch.Elapsed < maxTime)
                            {
                                var items = DummyLog2.Get();                                
                                
                                foreach (var item in items)
                                {
                                    itemsSet.Add(item);                                    
                                }
                                if (WaitHandle.WaitAny(waitHandles) == 1)
                                {
                                    break;
                                }
                            }

                            // Delete the consumed items
                            foreach (var item in itemsSet)
                            {
                                WriteLog("[Attempt{0}].Read: {1}", attempt, item);
                                item.Delete();
                            }
                        });
                }
            }
        }

        private static void WriteLog(string msg, params object[] args)
        {
            Console.WriteLine(msg, args);
            Debug.WriteLine(msg, args);
        }
    }

    [Database]
    public class DummyHead
    {
        private long m_sortId;

        public long SortId { get { return m_sortId; } }

        public long GetSortId()
        {
            return m_sortId++;
        }

        public static IEnumerable<DummyHead> Get()
        {
            return Db.SQL<DummyHead>("SELECT l FROM DummyHead l");
        }

        internal static void Purge()
        {
            foreach (var item in Get())
            {
                item.Delete();
            }
        }
    }

    [Database]
    public class DummyLog2
    {        
        public readonly long SortId;
        public readonly DateTime Created;
        public readonly string Message;
        public readonly DummyHead Head;

        public DummyLog2(DummyHead head, string msg)
        {
            Head = head;
            SortId = head.GetSortId();
            Created = DateTime.Now;
            Message = msg;
        }

        public static IEnumerable<DummyLog2> Get()
        {
            return Db.SQL<DummyLog2>("SELECT l FROM DummyLog2 l Order By l.Created");
        }

        internal static void Purge()
        {
            foreach (var item in Get())
            {
                item.Delete();
            }
        }

        public override string ToString()
        {
            return string.Format("[{0}] {1}", Created, Message);
        }
    }

    class DebugMonitor
    {
        private readonly System.Threading.CancellationTokenSource s_source;
        private readonly Timer m_timer;

        public DebugMonitor(System.Threading.CancellationTokenSource s_source)
        {
            if (Debugger.IsAttached)
            {
                Debug.WriteLine("Monitoring debugger");
                m_timer = new Timer(OnCheck, null, TimeSpan.FromSeconds(1), Timeout.InfiniteTimeSpan);
            }
        }

        private void OnCheck(object state)
        {
            if (!Debugger.IsAttached)
            {
                s_source.Cancel();
                m_timer.Dispose();
            }
            else
            {
                m_timer.Change(TimeSpan.FromSeconds(1), Timeout.InfiniteTimeSpan);
            }
        }

    }
}
```

**Comment by per-samuelsson** [comment]
Wow, what a nice sample to reproduce from. Thanks for the effort! Lets see where we get on this.

> I then have a consumer that reads (no write) item-records and once they are considered processed I delete them.

Not at all sure it matters for this case, but a *delete* would qualify as a modification. If you delete, you are no longer doing read-only processing.

**Comment by k-rus** [comment]
Consumer transaction will conflict with each other when they try to delete the same record. As I see in your code your transactions contain wait, which means that they will run too long and it will increase the probability of the conflict.
General practice is that snapshot isolated transactions, which guarantees consistency, should be short running. There should not be any waits, IOs and other expensive operations. For example, you can split your transaction into a read only part with read long running transaction (Scope), and a part, which does consistent changes.
You also need to think how to reduce the chance of the conflict between your threads/tasks.

In general you described an interesting pattern: a transactional queue.

**Comment by careri** [comment]
There is only one consumer thread. It should not conflict with any of the provider threads. 
The item has been written by one provider and commited. 
The commit hook then triggers an available item. Which the consumer will consume. The item didn't exist when the Db.Transact was created, is that the problem? That I delete an item that didn't exist when I started my Db.Transact?

**Comment by careri** [comment]
This is wrong
> Consumer transaction will conflict with each other when they try to delete the same record

There is only one consumer thread and therefore only one consumer transaction

**Comment by k-rus** [comment]
> That I delete an item that didn't exist when I started my Db.Transact?

I think it should be the problem. @eriohl can you confirm?

**Comment by careri** [comment]
I hope that is so, then I understand and can solve the problem.

**Comment by k-rus** [comment]
From theory: snapshot isolated transaction takes a snapshot of the database at the moment of starting transaction. Then if the transaction tries to write/modify data, which was committed by another transaction after the snapshot was taken, conflict happens. This is exactly the case here, however I don't know implementation details in Starcounter, thus it would be nice to get the confirmation.

**Comment by careri** [comment]
The defaults for Db.Transact indicates that snapshot isolation isn't used?
```public static void Transact(Action action, bool forceSnapshot = false, int maxRetries = 100);```

**Comment by k-rus** [comment]
> The defaults for Db.Transact indicates that snapshot isolation isn't used?
> public static void Transact(Action action, bool forceSnapshot = false, int maxRetries = 100);

No, it does not indicate. The flag means to force snapshot isolation even if the transaction takes too long, i.e., not allow switching from snapshot to read committed isolation level during executing transaction.
`Db.Transact` means that the transaction is started as snapshot isolated. `Db.Scope` means read committed.

**Comment by careri** [comment]
Ok. I'm not quite following. If it is snapshot as you say. Would it still find the new records commited by the provider thread?

**Comment by k-rus** [comment]
> Would it still find the new records commited by the provider thread?

Yes, I think so. Snapshot is guaranteed for writes. If I understand correctly, you can read data after they were committed in another thread. It is better to ask @eriohl for details. I feel I speculate here.

**Comment by eriohl** [comment]
There is a conflict between the provider and consumer then right? I would guess that the conflict is, as @k-rus suggests, that the consumer thread deletes a record created after it started.

Just to be clear here: The consumer thread will not have snapshot isolation because it waits. But conflict is based on snapshot isolation meaning that you update a transaction created after the transaction started. (This is not working properly obviously, has to do with this implicit fallback to a long running transaction.)

If it actually was snapshot isolation then there would not be a conflict because the consumer would not get any additional records after wait, regardless of how many the provider had created.

**Comment by careri** [comment]
Great, something I can understand :)
So, to summarize:
> snapshot = don't read any data modifed/added/deleted by other transactions

Right?

This issue helped my improve the design and security of my app. Thanks!

**Comment by eriohl** [comment]
Right: In snapshot isolation you would not see any updates provided by other transactions after the snapshot transaction started. Conflict occurs if data updated by the snapshot was updated after the transaction started (and snapshot was created).

But there is no practical way to ensure snapshot isolation because transaction can implicitly fall back to being a long running transaction. But it keeps the conflict detection, which is kind of inconsistent.

I'm guessing that you want to do some I/O or something in the consumer thread? Then should use a long running transaction there (which the conflict resolution disabled).

**Comment by careri** [comment]
I changed my code so that it looks somewhat like this (might be syntax error, texteditor coding):
```
// Create a batch  
var itemsSet = new HashSet<DummyLog2>();
var stopWatch = Stopwatch.StartNew();

while (itemsSet.Count < maxItems && stopWatch.Elapsed < maxTime)
{
  Db.Transact(() =>
  {
    var items = DummyLog2.Get();                                

    foreach (var item in items)
    {
        itemsSet.Add(item);                                    
    }
  });
  if (WaitHandle.WaitAny(waitHandles) == 1)
  {
      break;
  }     
}

// Delete the consumed items
Db.Transact(() =>
{  
  attempt++;
  foreach (var item in itemsSet)
  {
      WriteLog("[Attempt{0}].Read: {1}", attempt, item);
      item.Delete();
  }    
});
```

**Comment by k-rus** [comment]
> // Delete the consumed items
> Db.Transact(() =>
> {  
>   attempt++;
>   foreach (var item in itemsSet)
>   {
>       WriteLog("[Attempt{0}].Read: {1}", attempt, item);
>       item.Delete();
>   }    
> });

What is the reason for `WriteLog` call? Does it write to disk? If so, then it is expensive and should be moved out of the snapshot isolated scope. Furthermore, as you already noticed, it is called for every attempt.

**Comment by careri** [comment]
No it's just logging to Debug/Console used for debugging. And this isn't my real code.
